### ISSUE TITLE
AttributeError on Page edit when using ImageBlock

### ISSUE THREAD

**Issue by kaij** [wagtail/wagtail#12688]
### Issue Summary
We are experiencing a fatal error when trying to edit a page in the backend. It seems somehow related to using the new `ImageBlock` (Wagtail version: 6.3.1), but not sure how and why it came to this. There might be configuration error on our side or having to do with the custom image model. 

Since the ImageBlock is not working with wagtail_localize, it first seemed related to that. In the end, to remedy, I wrote a custom migration to migrate the streamfield back to ImageChooserBlock. After that, editing the page works flawless again. So it somehow does seem to be related to the nesting of ImageBlock in our StreamField...

<details><summary>Traceback</summary>

```
Django Version: 5.1.3
Python Version: 3.13.1
Installed Applications:
['django_vite',
 'csp',
 'corsheaders',
 'app.search',
 'app.tags',
 'app.image_annotations',
 'app.collection',
 'app.quizzes',
 'app.main',
 'app.content',
 'app.apps.LazyImagesAppConfig',
 'app',
 'wagtail_localize',
 'wagtail_localize.locales',
 'wagtail.contrib.forms',
 'wagtail.contrib.redirects',
 'wagtail.embeds',
 'wagtail.sites',
 'wagtail.users',
 'wagtail.snippets',
 'wagtail.documents',
 'wagtail.search',
 'wagtail.contrib.search_promotions',
 'wagtail',
 'wagtail.admin',
 'wagtail.api.v2',
 'django_svelte_jsoneditor',
 'wagtailmedia',
 'modelcluster',
 'taggit',
 'rest_framework',
 'django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.sitemaps',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'whitenoise.runserver_nostatic',
 'django.contrib.staticfiles',
 'dbbackup',
 'wagtailautocomplete',
 'wagtail_2fa',
 'django_otp',
 'django_otp.plugins.otp_totp',
 'wagtail_headless_preview']
Installed Middleware:
['django.middleware.security.SecurityMiddleware',
 'whitenoise.middleware.WhiteNoiseMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'corsheaders.middleware.CorsMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'csp.middleware.CSPMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'wagtail_2fa.middleware.VerifyUserPermissionsMiddleware',
 'django.middleware.locale.LocaleMiddleware',
 'wagtail.contrib.redirects.middleware.RedirectMiddleware']



Traceback (most recent call last):
  File "/venv/lib/python3.13/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.13/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.13/site-packages/django/views/decorators/cache.py", line 80, in _view_wrapper
    response = view_func(request, *args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.13/site-packages/wagtail/admin/urls/__init__.py", line 172, in wrapper
    return view_func(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.13/site-packages/wagtail/admin/auth.py", line 151, in decorated_view
    response = view_func(request, *args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.13/site-packages/django/views/generic/base.py", line 104, in view
    return self.dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.13/site-packages/wagtail/admin/views/pages/edit.py", line 396, in dispatch
    return super().dispatch(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.13/site-packages/django/views/generic/base.py", line 143, in dispatch
    return handler(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.13/site-packages/wagtail/admin/views/pages/edit.py", line 432, in get
    self.form = self.form_class(
                
  File "/venv/lib/python3.13/site-packages/wagtail/admin/forms/pages.py", line 175, in __init__
    super().__init__(data, files, *args, initial=initial, **kwargs)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.13/site-packages/wagtail/admin/forms/models.py", line 126, in __init__
    super().__init__(*args, **kwargs)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.13/site-packages/permissionedforms/forms.py", line 67, in __init__
    super().__init__(*args, **kwargs)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.13/site-packages/modelcluster/forms.py", line 334, in __init__
    self.formsets[rel_name] = formset_class(
                              
  File "/venv/lib/python3.13/site-packages/wagtail/admin/forms/comments.py", line 85, in __init__
    if comment.has_valid_contentpath(self.instance)
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.13/site-packages/wagtail/models/__init__.py", line 4797, in has_valid_contentpath
    block = field.get_block_by_content_path(stream_value, remainder)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.13/site-packages/wagtail/fields.py", line 272, in get_block_by_content_path
    return self.stream_block.get_block_by_content_path(value, path_elements)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.13/site-packages/wagtail/blocks/stream_block.py", line 423, in get_block_by_content_path
    return child.block.get_block_by_content_path(
           
  File "/venv/lib/python3.13/site-packages/wagtail/blocks/struct_block.py", line 302, in get_block_by_content_path
    return child_block.get_block_by_content_path(
           
  File "/venv/lib/python3.13/site-packages/wagtail/blocks/stream_block.py", line 423, in get_block_by_content_path
    return child.block.get_block_by_content_path(
           
  File "/venv/lib/python3.13/site-packages/wagtail/blocks/struct_block.py", line 302, in get_block_by_content_path
    return child_block.get_block_by_content_path(
           
  File "/venv/lib/python3.13/site-packages/wagtail/blocks/struct_block.py", line 301, in get_block_by_content_path
    child_value = value.get(name, child_block.get_default())
                  ^^^^^^^^^

Exception Type: AttributeError at /backend/admin/pages/351/edit/
Exception Value: 'CustomImage' object has no attribute 'get'
```

</details>

**Comment by gasman** [comment]
Thanks for the report @kaij!

From the traceback, it looks like ImageBlock is lacking an override for StructBlock's `get_block_by_content_path` method, which is used by the commenting mechanism - so it's likely that this bug surfaces specifically when a user has commented on a sub-block of the ImageBlock.

(It's possible that wagtail_localize makes use of the same method too.)

**Comment by zerolab** [comment]
wagtail-localize fails to extract segments off the individual blocks in https://github.com/wagtail/wagtail-localize/blob/main/wagtail_localize/segments/extract.py#L123-L140 as `ImageBlock` is a `StructBlock` subclass, but unlike a `StructBlock` it deconstucts differently https://github.com/wagtail/wagtail/blob/9889bddcb8823369f7a26c6b6f4a3eb6784d5985/wagtail/images/blocks.py#L104 and the block value ends up being the image.

**Comment by Chadys** [comment]
I created a dedicated issue on Wagtail-localize side. This issue might need both repo collaboration to be fixed. For now, I am going back to using `ImageChooserBlock`

**Comment by kaij** [comment]
> From the traceback, it looks like ImageBlock is lacking an override for StructBlock's `get_block_by_content_path` method, which is used by the commenting mechanism - so it's likely that this bug surfaces specifically when a user has commented on a sub-block of the ImageBlock.

Thanks @gasman. I checked and see that there are indeed comments on the blocks. So I guess this is a separate issue apart from wagtail-localize.

**Comment by gasman** [comment]
Have replicated on [bakerydemo](https://github.com/wagtail/bakerydemo/) as follows:

* Add an ImageBlock to `bakerydemo.base.blocks.BaseStreamBlock`: in blocks.py, add `from wagtail.images.blocks import ImageBlock as WagtailImageBlock`, and append `better_image = WagtailImageBlock()` to `class BaseStreamBlock`
* Edit a blog page; insert a "better image" block, select an image and fill in alt text, then use the comment button on the image field to add a comment
* Save as draft; receive error as above.

Fix now submitted at #12702.
